Introduce fromEgressProxyRule

jschwinger233 opened the pull request:
This simple PR introduces a new ip rule without installing it, mainly meant to take care of cilium downgrade for the new ip rule. Please see the commit messages for details.

This PR (along with its backport PRs) is the stage one of #31984 to address the remaining IPsec leak issues. Please see that issue for the the context.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
proxy/routes: Rename fromProxyRule to fromIngressProxyRule

Because we are introducing fromEgressProxyRule soon, it's better to make
clear that the fromProxyRule is for ingress proxy only.

This commit also changes its mark from MagicMarkIsProxy to
MagicMarkIngress. They hold the same value 0xA00 while have the
different semantics.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
proxy/routes: Introduce fromEgressProxyRule

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
proxy/routes: Remove fromEgressProxyRule for cilium downgrade

Although we don't install fromEgressProxyRule for now, this commit
insists on removing it to make sure further downgrade can go smoothly.

Soon We'll have another PR to install fromEgressProxyRule, and cilium
downgrade from that PR to branch tip (patch downgrade, 1.X.Y ->
1.X.{Y-1}) will be broken if we don't handle the new ip rule carefullly.

Without this patch, downgrade from higher version will leave
fromEgressProxyRule on the lower version cilium, cluster will be in a
wrong status of "having stale ip rule + not having other necessary
settings (iptables)", breaking the connectivity.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

rgo3 commented on pkg/proxy/routes.go:
```diff
@@ -100,10 +100,19 @@ func removeToProxyRoutesIPv6() error {
+
+	//nolint:unused // Routing rule for traffic from egress proxy.
+	fromEgressProxyRule = route.Rule{
+		Priority: linux_defaults.RulePriorityFromProxyIngress,
```
Should this be the Ingress rule priority? From a quick glance I see we also have priority settings for egress rules

jschwinger233 committed:
proxy/routes: Rename RulePriorityFromProxyIngress

No logic changes, just rename it to "RulePriorityFromProxy" without
"Ingress" suffix, because egress rule is using the same priority.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

julianwiedmann commented on pkg/proxy/routes.go:
```diff
@@ -100,10 +100,19 @@ func removeToProxyRoutesIPv6() error {
+		Protocol: linux_defaults.RTProto,
+	}
+
+	//nolint:unused // Routing rule for traffic from egress proxy.
```
```suggestion
	//nolint:unused // Routing rule for traffic from egress proxy.
```

huh, why is the `unused` needed? We *are* using it below, no?

jschwinger233 replied:
Because it's in a separate commit. Should I make them together?

julianwiedmann replied:
ah! No, then I think it's ok, but the last commit should remove the `nolint` again?

julianwiedmann commented on pkg/proxy/proxy.go:
```diff
@@ -411,19 +411,22 @@ func (p *Proxy) ReinstallRoutingRules() error {
 		}
+		if err := removeFromEgressProxyRoutesIPv4(); err != nil {
+			return err
+		}
```
This feels a bit awkward, as `removeFromProxyRoutesIPv4()` already deletes the whole `RouteTableFromProxy` table. 

Should we maybe put the call to `removeFromEgressProxyRoutesIPv4()` *prior* to the `if !option.Config.EnableIPSec || option.Config.TunnelingEnabled()` handling?

jschwinger233 replied:
Sure!

julianwiedmann reviewed (commented):
ty! Sorry for taking so long - two inline comments, depends on your response whether they should be "request changes" :slightly_smiling_face:

julianwiedmann reviewed (approved):
lgtm, thank you!

tommyp1ckles commented on pkg/proxy/proxy.go:
```diff
@@ -410,6 +410,9 @@ func (p *Proxy) ReinstallRoutingRules() error {
 			return err
 		}
 
+		if err := removeFromEgressProxyRoutesIPv4(); err != nil {
```
Does this mean that this change, and the subsequent one to add the egress proxy rules must be in introduced in different patch versions?

jschwinger233 replied:
Yes, will be backported to all stable branches.
